File: tutorials/mpi/examples/ser_mm.c
/******************************************************************************
* FILE: ser_mm.c
* DESCRIPTION:  
*   Serial Matrix Multiply - C Version
* AUTHOR: Blaise Barney
* LAST REVISED: 04/12/05
******************************************************************************/
#include <stdio.h>
#include <stdlib.h>

#define NRA 62 			/* number of rows in matrix A */
#define NCA 15			/* number of columns in matrix A */
#define NCB 7   		/* number of columns in matrix B */

int main(int argc, char *argv[])
{
int    i, j, k;			/* misc */
double a[NRA][NCA], 		/* matrix A to be multiplied */
       b[NCA][NCB],      	/* matrix B to be multiplied */
       c[NRA][NCB];		/* result matrix C */

printf("Starting serial matrix multiple example...\n");
printf("Using matrix sizes a[%d][%d], b[%d][%d], c[%d][%d]\n",
        NRA,NCA,NCA,NCB,NRA,NCB);

/* Initialize A, B, and C matrices */
printf("Initializing matrices...\n");
for (i=0; i<NRA; i++)
   for (j=0; j<NCA; j++)
      a[i][j]= i+j;
for (i=0; i<NCA; i++)
   for (j=0; j<NCB; j++)
      b[i][j]= i*j;
for(i=0;i<NRA;i++)
   for(j=0;j<NCB;j++)
      c[i][j] = 0.0;

/* Perform matrix multiply */
printf("Performing matrix multiply...\n");
for(i=0;i<NRA;i++)
   for(j=0;j<NCB;j++)
      for(k=0;k<NCA;k++)
         c[i][j]+= a[i][k] * b[k][j];

printf("Here is the result matrix:");
for (i=0; i<NRA; i++) { 
   printf("\n"); 
   for (j=0; j<NCB; j++) 
      printf("%6.2f   ", c[i][j]);
   }
printf ("\nDone.\n");
}

File: mpi/examples/ser_mm.f
C******************************************************************************
C FILE: ser_mm.f
C DESCRIPTION: 
C   Serial Matrix Multiply - Fortran version.
C AUTHOR: unknown
C LAST REVISED:  04/15/05 Blaise Barney 
C******************************************************************************

      program mm

      parameter (NRA = 62)
      parameter (NCA = 15)
      parameter (NCB = 7)

      integer 	i, j, k
      real*8	a(NRA,NCA), b(NCA,NCB), c(NRA,NCB)

      print *,'Starting serial matrix multiple example...'
      write(*,100) NRA,NCA,NCA,NCB,NRA,NCB
 100  format(' Using matrix sizes a(',I2,',',I2,'),b(',I2,',',I1,
     & '),c(',I2,',',I1,')')

C     Initialize A and B 
      print *,'Initializing matrices...'
      do 30 i=1, NRA
        do 30 j=1, NCA
        a(i,j) = (i-1)+(j-1)
 30   continue
      do 40 i=1, NCA
        do 40 j=1, NCB
	b(i,j) = (i-1)*(j-1)
 40   continue

C     Do matrix multiply
      print *,'Performing matrix multiply...'
       do 50 k=1, NCB 
         do 50 i=1, NRA
         c(i,k) = 0.0
           do 50 j=1, NCA
           c(i,k) = c(i,k) + a(i,j) * b(j,k)
  50       continue

C     Print results 
      print*, 'Here is the result matrix: '	
      do 90 i=1, NRA
         do 80 j = 1, NCB
         write(*,70)c(i,j)
  70     format(2x,f8.2,$)
  80     continue
      print *, ' '
  90  continue
      print *, 'Done.'
      print *, ' '

      end
